From e5823fa953975b34336f1c054027bbd925a7b81e Mon Sep 17 00:00:00 2001
From: Yannick Fertre <yannick.fertre@foss.st.com>
Date: Tue, 1 Apr 2025 17:45:58 +0200
Subject: [PATCH 60/95] drm/stm: dsi: add callback
 dw_mipi_dsi_stm_phy_141_mode_valid

This callback is specific to physical layer 141 & linked to
the PLL calculation function.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/linux.git 8541b22aac38d79c22368275f61bf57f106a8c68 drm/stm: dsi: add callback dw_mipi_dsi_stm_phy_141_mode_valid]
Change-Id: I56b8dfaaf4b14333c0ff254228688fc6ac4c75a9
Signed-off-by: Yannick Fertre <yannick.fertre@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/451888
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/gpu/drm/stm/dw_mipi_dsi-stm.c | 127 +++++++++++++++++++++++---
 1 file changed, 113 insertions(+), 14 deletions(-)

diff --git a/drivers/gpu/drm/stm/dw_mipi_dsi-stm.c b/drivers/gpu/drm/stm/dw_mipi_dsi-stm.c
index cca890cfec39c..124451dfb1562 100644
--- a/drivers/gpu/drm/stm/dw_mipi_dsi-stm.c
+++ b/drivers/gpu/drm/stm/dw_mipi_dsi-stm.c
@@ -1101,20 +1101,119 @@ dw_mipi_dsi_stm_mode_valid(void *priv_data,
 
 		lane_mbps = pll_out_khz / 1000;
 
-		if (dsi->hw_version < HWVER_141) {
-			ret = dw_mipi_dsi_phy_get_timing(priv_data, lane_mbps, &dphy_timing);
-			if (ret)
-				return MODE_ERROR;
-			/*
-			 * In non-burst mode DSI has to enter in LP during HFP
-			 * (horizontal front porch) or HBP (horizontal back porch) to
-			 * resync with LTDC pixel clock.
-			 */
-			delay_to_lp = DIV_ROUND_UP((dphy_timing.data_hs2lp + dphy_timing.data_lp2hs)
-						   * lanes * BITS_PER_BYTE, bpp);
-			if (hfp < delay_to_lp && hbp < delay_to_lp)
-				return MODE_HSYNC;
+		ret = dw_mipi_dsi_phy_get_timing(priv_data, lane_mbps, &dphy_timing);
+		if (ret)
+			return MODE_ERROR;
+		/*
+		 * In non-burst mode DSI has to enter in LP during HFP
+		 * (horizontal front porch) or HBP (horizontal back porch) to
+		 * resync with LTDC pixel clock.
+		 */
+		delay_to_lp = DIV_ROUND_UP((dphy_timing.data_hs2lp + dphy_timing.data_lp2hs)
+					   * lanes * BITS_PER_BYTE, bpp);
+		if (hfp < delay_to_lp && hbp < delay_to_lp)
+			return MODE_HSYNC;
+	}
+
+	return MODE_OK;
+}
+
+static enum drm_mode_status
+dw_mipi_dsi_stm_phy_141_mode_valid(void *priv_data,
+				   const struct drm_display_mode *mode,
+				   unsigned long mode_flags, u32 lanes, u32 format)
+{
+	struct dw_mipi_dsi_stm *dsi = priv_data;
+	unsigned int lane_kbps;
+	int bpp;
+
+	bpp = mipi_dsi_pixel_format_to_bpp(format);
+	if (bpp < 0)
+		return MODE_BAD;
+
+	/* Compute requested lane data rate */
+	lane_kbps = mode->clock * bpp / lanes;
+
+	/* Add 20% to lane data rate to be higher than pixel bw */
+	if (mode_flags & MIPI_DSI_MODE_VIDEO_BURST)
+		lane_kbps = (lane_kbps * 12) / 10;
+
+	if (lane_kbps > dsi->lane_max_kbps)
+		return MODE_CLOCK_HIGH;
+
+	if (lane_kbps < dsi->lane_min_kbps)
+		return MODE_CLOCK_LOW;
+
+	if (!(mode_flags & MIPI_DSI_MODE_VIDEO_BURST)) {
+		unsigned int idf, ndiv, odf, pll_in_khz, pll_out_khz;
+		unsigned int px_clock_hz, target_px_clock_min, target_px_clock_max;
+		int dsi_short_packet_size_px, hfp, hsync, hbp, delay_to_lp, ret;
+		struct dw_mipi_dsi_dphy_timing dphy_timing;
+
+		/* Compute best pll parameters */
+		idf = 0;
+		ndiv = 0;
+		odf = 0;
+		pll_in_khz = clk_get_rate(dsi->pllref_clk) / 1000;
+		ret = dsi_phy_141_pll_get_params(dsi, pll_in_khz, lane_kbps / 2,
+						 &idf, &ndiv, &odf, NULL);
+		if (ret) {
+			DRM_WARN("Warning dsi_phy_141_pll_get_params(): bad params\n");
+			return MODE_ERROR;
+		}
+
+		/* Get the adjusted pll out value */
+		pll_out_khz = dsi_pll_get_clkout_khz(pll_in_khz, idf, ndiv, odf);
+
+		px_clock_hz = DIV_ROUND_CLOSEST_ULL(1000ULL * pll_out_khz * lanes * 2, bpp);
+
+		/*
+		 * Filter modes according to the clock value, particularly useful for
+		 * hdmi modes that require precise pixel clocks.
+		 */
+		target_px_clock_min = mode->clock * (1000 - clock_tolerance);
+		target_px_clock_max = mode->clock * (1000 + clock_tolerance);
+
+		if (px_clock_hz < target_px_clock_min)
+			return MODE_CLOCK_LOW;
+
+		if (px_clock_hz > target_px_clock_max)
+			return MODE_CLOCK_HIGH;
+
+		/* sync packets are codes as DSI short packets (4 bytes) */
+		dsi_short_packet_size_px = DIV_ROUND_UP(4 * BITS_PER_BYTE, bpp);
+
+		hfp = mode->hsync_start - mode->hdisplay;
+		hsync = mode->hsync_end - mode->hsync_start;
+		hbp = mode->htotal - mode->hsync_end;
+
+		/* hsync must be longer than 4 bytes HSS packets */
+		if (hsync < dsi_short_packet_size_px)
+			return MODE_HSYNC_NARROW;
+
+		if (mode_flags & MIPI_DSI_MODE_VIDEO_SYNC_PULSE) {
+			/* HBP must be longer than 4 bytes HSE packets */
+			if (hbp < dsi_short_packet_size_px)
+				return MODE_HSYNC_NARROW;
+			hbp -= dsi_short_packet_size_px;
+		} else {
+			/* With sync events HBP extends in the hsync */
+			hbp += hsync - dsi_short_packet_size_px;
 		}
+
+		ret = dw_mipi_dsi_phy_141_get_timing(priv_data, lane_kbps / 1000, &dphy_timing);
+		if (ret)
+			return MODE_ERROR;
+		/*
+		 * In non-burst mode DSI has to enter in LP during HFP
+		 * (horizontal front porch) or HBP (horizontal back porch) to
+		 * resync with LTDC pixel clock.
+		 */
+		delay_to_lp = DIV_ROUND_UP((dphy_timing.data_hs2lp + dphy_timing.data_lp2hs)
+					   * lanes * BITS_PER_BYTE / 2, bpp);
+
+		if (hfp < delay_to_lp && hbp < delay_to_lp)
+			return MODE_HSYNC;
 	}
 
 	return MODE_OK;
@@ -1144,7 +1243,7 @@ static struct dw_mipi_dsi_plat_data dw_mipi_dsi_stm_plat_data = {
 
 static struct dw_mipi_dsi_plat_data dw_mipi_dsi_stm32mp25_plat_data = {
 	.max_data_lanes = 4,
-	.mode_valid = dw_mipi_dsi_stm_mode_valid,
+	.mode_valid = dw_mipi_dsi_stm_phy_141_mode_valid,
 	.phy_ops = &dw_mipi_dsi_stm_phy_141_ops,
 };
 
-- 
2.51.0

